docs/deploying.adoc

```diff
@@ -9,11 +9,242 @@ Put the filter jars in some directory
+
+=== Cluster Network Address Config Providers
+
+The Cluster Network Address Config Provider controls how the virtual cluster is presented oo the network. Two
+alternatives are supported: `PortPerBroker` and `SniRouting` which have different characteristics which make each
+suitable for different use-cases. They are described next.
+
+==== PortPerBroker scheme
+
+In the `PortPerBroker` scheme, Kroxylicious automatically opens port for each virtual cluster's bootstrap and
+one port per broker of each target cluster.  So, if you have two virtual clusters, each targeting a Kafka Cluster
+of three brokers, Kroxylicious will bind eight ports in total.
+
+The `PortPerBroker` scheme can be used with either upstream plain or upstream TLS connections.
+
+[source, yaml]
+----
+clusterNetworkAddressConfigProvider:
+  type: PortPerBroker
+  config:
+    bootstrapAddress: mycluster.kafka.com:9192                                 # <1>
+    brokerAddressPattern: mybroker-$(nodeId).mycluster.kafka.com:$(portNumber) # <2>
+    brokerStartPort: 9193                                                      # <3>
+    numberOfBrokerPorts: 3                                                     # <4>
+    bindAddress: 192.168.0.1                                                   # <5>
+----
+<1> The hostname and port of the bootstrap that will be used by the Kafka Clients.
+<2> (Optional) The broker address pattern used to form the broker addresses.  If not defined, it defaults to the
+hostname part of the `bootstrapAddress` and the port number allocated to the broker.
+<3> (Optional) The starting number for broker port range. Defaults to the port of the `bootstrapAddress` plus 1.
+<4> (Optional) The maximum number of brokers of ports that will be defined from the range.  Defaults to 3.
+<5> (Optional) The bind address used when binding the ports. If undefined, all network interfaces will be bound.
+
+The `brokerAddressPattern` configuration parameters understands two replacement tokens.
+
+* this `$(portNumber)` is mandatory.  It is replaced by the port number computed for this broker.
+* the `$(nodeId)` token is optional.  If present, it will be replaced by the
+https://kafka.apache.org/documentation/#brokerconfigs_node.id[`node.id`] (or `broker.id`) assigned to the broker
+of the target cluster.
+
+For example if your configuration looks like the above and your cluster has three brokers, your Kafka Client will receive
+broker address information like this:
+
+0.  `mybroker-0.mycluster.kafka.com:9193`
+1.  `mybroker-1.mycluster.kafka.com:9194`
+2.  `mybroker-2.mycluster.kafka.com:9194`
+
+NOTE: It is a responsible for the deployer of Kroxylicious to ensure that the bootstrap and generated broker
+DNS names are resolvable and routable by the Kafka Client.
+
+==== SniRouting scheme
+
+In the `SniRouting` scheme, Kroxylicious uses SNI information to route traffic to either the boostrap or individual
+brokers.   As this relies on SNI, the use of upstream TLS is *required*.
```

Comment:
Could link to the TLS section here, also is it supposed to be Downstream TLS? The client is connecting to kroxylicious with TLS/SNI.

Reply:
Done.